File: sorting/mergesort.clj
(defn merge [right left]
  (cond
    (empty? left) right
    (empty? right) left
    :else
    (if (< (first left) (first right))
      (cons (first left) (merge (rest left) right))
      (cons (first right) (merge left (rest right))))))

(defn mergesort [list]
  (if (empty? (rest list))
   list
   (let [halves (split-at (/ (count list) 2) list)
         left (halves 0)
         right (halves 1)]
   (merge (mergesort left) (mergesort right)))))

(defn main []
  (let [list (shuffle '(0 1 2 3 4 5 6 7 8 9))
        sorted-list (mergesort list)]
    (if (= sorted-list (sort list))
      sorted-list
      (cons "error" sorted-list))))


(println (main))

File: sorting/mergesort.pl
merge(Left, [], Left):- !.
merge([], Right, Right):- !.
merge([LH|LT], [RH|RT], NewList):-
    (LH < RH; LH =:= RH),
    merge(LT, [RH|RT], Rest),
    !,
    append([LH], Rest, NewList).
merge([LH|LT], [RH|RT], NewList):-
    RH < LH,
    merge([LH|LT], RT, Rest),
    !,
    append([RH], Rest, NewList).


split_list(Xs, Ys, Zs) :-
    length(Xs, N),
    H is N - N // 2,
    length(Ys, H),
    append(Ys, Zs, Xs).

mergesort([], []):- !.
mergesort([X], [X]):- !.
mergesort(List, Sorted):-
    split_list(List, Left, Right),
    mergesort(Left, SortedLeft),
    mergesort(Right, SortedRight),
    !,
    merge(SortedLeft, SortedRight, Sorted).
